src/os/bluestore/BlueFS.cc

```diff
@@ -3478,11 +3695,25 @@ int BlueFS::_flush_range_F(FileWriter *h, uint64_t offset, uint64_t length)
     }
     h->file->is_dirty = true;
   }
-  if (h->file->fnode.size < offset + length) {
-    vselector->add_usage(h->file->vselector_hint, offset + length - h->file->fnode.size);
-    h->file->fnode.size = offset + length;
+  if (h->file->fnode.size < end) {
+    vselector->add_usage(h->file->vselector_hint, end - h->file->fnode.size);
+    h->file->fnode.size = end;
     h->file->is_dirty = true;
   }
+
+  if (h->file->is_new_wal()) {
+    // create WAL flush envelope
+    uint64_t flush_size = length - File::WALFlush::extra_envelope_size_on_front_and_tail();
```

Comment:
IMO one can define flush_size as ceph_le64 flush_size_le at this point and avoid further tossing